### PR TITLE
Forwarder storage

### PR DESCRIPTION
**Pull request by KodrAus**
This PR introduces a durable WAL for the forwarder to ingest events into. It's a pretty sizable piece of software even though it's simple so deserves some documentation once we're ready to stabilize the whole thing.

### CODE REVIEW

File: src/SeqCli/Forwarder/Storage/BufferAppender.cs [hunk 1]
```diff
@@ -0,0 +1,164 @@
+// Copyright Datalust Pty Ltd
```
Comment (nblumhardt) [inline src/SeqCli/Forwarder/Storage/BufferAppender.cs:1]: Missing the © in these headers (consistency with the others)

File: src/SeqCli/Forwarder/Storage/BufferAppender.cs [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    /// <param name="maxChunks">The maximum number of chunk files to keep before starting to delete them. This
+    /// is an optional parameter to use in cases where the reader isn't keeping up with the writer.</param>
+    /// <param name="sync">Whether to explicitly flush the write to disk.</param>
+    /// <returns>True if the write fully succeeded. If this method returns false, it is safe to retry the write,
+    /// but it may result in duplicate data in the case of partial success.</returns>
+    public bool TryAppend(Span<byte> batch, long targetChunkLength, int? maxChunks = null, bool sync = true)
+    {
+        if (batch.Length == 0) return true;
+
+        if (batch[^1] != (byte)'\n') throw new Exception("Batches must end with a newline character (\\n)");
```
Comment (nblumhardt) [inline src/SeqCli/Forwarder/Storage/BufferAppender.cs:72]: Exception message should end in `.`

File: src/SeqCli/Forwarder/Storage/BufferAppender.cs [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+        if (_currentChunk != null)
```
Comment (nblumhardt) [inline src/SeqCli/Forwarder/Storage/BufferAppender.cs:74]: Multiline `if` generally gets `{` and `}` in this codebase

File: src/SeqCli/Forwarder/Storage/BufferAppender.cs [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+    public static BufferAppender Open(StoreDirectory storeDirectory)
+    {
+        return new BufferAppender(storeDirectory);
+    }
+
+    /// <summary>
+    /// Try write a batch.
+    ///
+    /// This method does not throw.
```
Comment (nblumhardt) [inline src/SeqCli/Forwarder/Storage/BufferAppender.cs:50]: Might be good to add more detail around this? Seems like `_currentChunk.Appender.Sync()` would/could throw `IOException`?

File: src/SeqCli/Forwarder/Storage/BufferAppenderChunk.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using SeqCli.Forwarder.Filesystem;
+
+namespace SeqCli.Forwarder.Storage;
+
+sealed record BufferAppenderChunk(StoreFileAppender Appender) : IDisposable
```
Comment (nblumhardt) [inline src/SeqCli/Forwarder/Storage/BufferAppenderChunk.cs:20]: Records are normally value-like; it'd be strange for the generated equality/hash implementation etc. on this type to include `Appender`. Should probably just be `class`?

File: src/SeqCli/Forwarder/Storage/BufferReaderChunk.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+using System;
+using SeqCli.Forwarder.Filesystem;
+
+namespace SeqCli.Forwarder.Storage;
+
+/// <summary>
+///     An active chunk in a <see cref="BufferReader" />.
+/// </summary>
+record BufferReaderChunk(ChunkName Name, StoreFile Chunk) : IDisposable
```
Comment (nblumhardt) [inline src/SeqCli/Forwarder/Storage/BufferReaderChunk.cs:23]: Maybe just `class`?

File: test/SeqCli.Tests/Forwarder/Filesystem/InMemoryStoreDirectory.cs [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+public class InMemoryStoreDirectory : StoreDirectory
+{
+    readonly Dictionary<string, InMemoryStoreFile> _files = new();
+
+    public IReadOnlyDictionary<string, InMemoryStoreFile> Files => _files;
+
+    public override InMemoryStoreFile Create(string name)
+    {
+        if (_files.ContainsKey(name)) throw new Exception($"The file {name} already exists");
```
Comment (nblumhardt) [inline test/SeqCli.Tests/Forwarder/Filesystem/InMemoryStoreDirectory.cs:16]: Trailing `.`?

File: src/SeqCli/Forwarder/Filesystem/System/SystemStoreDirectory.cs [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+        var dir = Libc.open(directoryPath, 0);
+        if (dir == -1) return;
+
+        // NOTE: directory syncing here is best-effort
+        // If it fails for any reason we simply carry on
+#pragma warning disable CA1806
+        Libc.fsync(dir);
+        Libc.close(dir);
+#pragma warning restore CA1806
```
Comment (nblumhardt) [inline src/SeqCli/Forwarder/Filesystem/System/SystemStoreDirectory.cs:123]: Should these be in `#if UNIX`?

**Review by nblumhardt — COMMENTED** [review]
Planning to spend some time with this in detail once merged, but LGTM otherwise :+1:

File: src/SeqCli/Forwarder/Storage/BufferAppender.cs (hunk 4 above)
Comment (KodrAus) [reply src/SeqCli/Forwarder/Storage/BufferAppender.cs:50]: We catch those

File: src/SeqCli/Forwarder/Filesystem/System/SystemStoreFileAppender.cs [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.IO;
+
+namespace SeqCli.Forwarder.Filesystem.System;
+
+public sealed class SystemStoreFileAppender : StoreFileAppender
```
Comment (nblumhardt) [inline src/SeqCli/Forwarder/Filesystem/System/SystemStoreFileAppender.cs:20]: Nit: few types in this project need to be `public`; default internal is fine

**Pull request merged**